### PR TITLE
Multiple delivery client factory

### LINKED ISSUE #312: TypeProvider is not registered properly when using RegisterType
### Brief bug description

When registering type provider with ` builder.RegisterType<TProvider>().Named<ITypeProvider>('x');`  the `ITypeProvider` is not actually registered and `GetType` method within provider is never called when client is fetched using `IDeliveryClientFactory` 

### Expected behavior

TypeProvider should be invoked when delivery client is initialized through `IDeliveryClientFactory` 

### Test environment

.NET 6, VS 2022, Kentico.Kontent.Delivery@16.0.0-beta5

### PR DESCRIPTION
**Pull request by Simply007**
### Motivation

Fixes #312 (replaced #333)

DEVREL-601

TODO

- [x] Adjust documentation

### Checklist

- [ ] Code follows coding conventions held in this repo
- [ ] Automated tests have been added
- [ ] Tests are passing
- [ ] Docs have been updated (if applicable)
- [ ] Temporary settings (e.g. variables used during development and testing) have been reverted to defaults

### How to test

If manual testing is required, what are the steps?

### CODE REVIEW

**Comment by Simply007** [comment]
One last question - `MultipleDleiveryClientFactory` is OK?

File: Kontent.Ai.Delivery.Extensions.DependencyInjection/Builders/MultipleDeliveryClientFactoryBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+            var client = BuildDeliveryClient(deliveryOptionsBuilder, optionalClientSetup);
+            RegisterClient(name, client);
+
+            return this;
+        }
+
+        public IMultipleDeliveryClientFactoryBuilder AddDeliveryClientCache
+        (
+            string name, Func<IDeliveryOptionsBuilder,
```
Comment (gormal) [inline Kontent.Ai.Delivery.Extensions.DependencyInjection/Builders/MultipleDeliveryClientFactoryBuilder.cs:41]: One parameter per line would be better.

File: Kontent.Ai.Delivery/Builders/DeliveryClient/IMultipleDeliveryClientFactoryBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+using System;
+using Kontent.Ai.Delivery.Abstractions;
+using Kontent.Ai.Delivery.Builders.DeliveryClient;
+using Kontent.Ai.Delivery.Configuration;
+
+namespace Kontent.Ai.Delivery.Builders
+{
+    // TODO 312 - XML comments
```
Comment (gormal) [inline Kontent.Ai.Delivery/Builders/DeliveryClient/IMultipleDeliveryClientFactoryBuilder.cs:8]: There are XML comments missing.

File: Kontent.Ai.Delivery/Builders/DeliveryClient/IMultipleDeliveryClientFactoryBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+using System;
+using Kontent.Ai.Delivery.Abstractions;
+using Kontent.Ai.Delivery.Builders.DeliveryClient;
+using Kontent.Ai.Delivery.Configuration;
+
+namespace Kontent.Ai.Delivery.Builders
+{
+    // TODO 312 - XML comments
+    // TODO 312 decide whether this lib is the best for this interface (currently here because of IDeliveryClientBuilder dependency)
```
Comment (gormal) [inline Kontent.Ai.Delivery/Builders/DeliveryClient/IMultipleDeliveryClientFactoryBuilder.cs:9]: Haven't we talked about this already?

**Review by gormal — CHANGES_REQUESTED** [review]
There are still some to-do comments and missing XML documentation.

File: Kontent.Ai.Delivery/Builders/DeliveryClient/IMultipleDeliveryClientFactoryBuilder.cs (hunk 2 above)
Comment (Simply007) [reply Kontent.Ai.Delivery/Builders/DeliveryClient/IMultipleDeliveryClientFactoryBuilder.cs]: 6cdbcc6

File: Kontent.Ai.Delivery.Extensions.DependencyInjection/Builders/MultipleDeliveryClientFactoryBuilder.cs (hunk 1 above)
Comment (Simply007) [reply Kontent.Ai.Delivery.Extensions.DependencyInjection/Builders/MultipleDeliveryClientFactoryBuilder.cs:41]: 6cdbcc6

File: Kontent.Ai.Delivery.Extensions.DependencyInjection/Builders/MultipleDeliveryClientFactoryBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+{
+    internal class MultipleDeliveryClientFactoryBuilder : IMultipleDeliveryClientFactoryBuilder
+    {
+        private readonly ConcurrentDictionary<string, IDeliveryClient> _clients = new();
+
+        public IMultipleDeliveryClientFactoryBuilder AddDeliveryClient
+        (
+            string name,
+            Func<IDeliveryOptionsBuilder,
+            DeliveryOptions> deliveryOptionsBuilder,
```
[the remark this answers is not in the document]
Comment (Simply007) [reply Kontent.Ai.Delivery.Extensions.DependencyInjection/Builders/MultipleDeliveryClientFactoryBuilder.cs:19]: 6cdbcc6

**Comment by Simply007** [comment]
Went through the comments -> is that alright @gormal?

File: Kontent.Ai.Delivery.Extensions.DependencyInjection/Builders/MultipleDeliveryClientFactoryBuilder.cs (hunk 1 above)
Comment (gormal) [reply Kontent.Ai.Delivery.Extensions.DependencyInjection/Builders/MultipleDeliveryClientFactoryBuilder.cs:41]: There are 4 parameters and 4 lines but still, it is not one parameter per line. 

```
            string name,
            Func<IDeliveryOptionsBuilder, DeliveryOptions> deliveryOptionsBuilder,
            IDeliveryCacheManager cacheManager,
            Func<IOptionalClientSetup, IOptionalClientSetup> optionalClientSetup = null
```

File: Kontent.Ai.Delivery.Extensions.DependencyInjection/Builders/MultipleDeliveryClientFactoryBuilder.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (gormal) [reply Kontent.Ai.Delivery.Extensions.DependencyInjection/Builders/MultipleDeliveryClientFactoryBuilder.cs:19]:
```
            string name,
            Func<IDeliveryOptionsBuilder, DeliveryOptions> deliveryOptionsBuilder,
            Func<IOptionalClientSetup, IOptionalClientSetup> optionalClientSetup = null
```

**Comment by gormal** [comment]
> Went through the comments -> is that alright @gormal?

Just nitpicks with formatting, please have a look.

**Review by gormal — APPROVED** [review]
LGTM!

**Review by dzmitryk-kontent-ai — CHANGES_REQUESTED** [review]

File: docs/configuration/multiple-delivery-clients.md [hunk 5]
```diff
@@ -44,13 +136,30 @@ public class Startup
 
     public void ConfigureServices(IServiceCollection services)
     {
-        services.AddDeliveryClient("projectA", Configuration, "ProjectAOptions", NamedServiceProviderType.Autofac);
-        services.AddDeliveryClient("projectB", Configuration, "ProjectBOptions", NamedServiceProviderType.Autofac);
+        services.AddDeliveryClientFactory
```
[the remark this answers is not in the document]
Comment (Simply007) [reply docs/configuration/multiple-delivery-clients.md]: [97b049e](https://github.com/kontent-ai/delivery-sdk-net/pull/347/commits/97b049e5fd2ffb706af424c14093eb80465e0d1d)

File: docs/configuration/multiple-delivery-clients.md [hunk 6]
```diff
@@ -1,36 +1,128 @@
...
+If you're accessing two completely different projects, chances are they have a different content model and therefore the generated models for content types will differ. Extend the Startup class as follows:
+
+```csharp
+public class Startup
+{
+    // ...
+
+    public void ConfigureServices(IServiceCollection services)
+    {
+        services.AddDeliveryClientFactory
```
[the remark this answers is not in the document]
Comment (Simply007) [reply docs/configuration/multiple-delivery-clients.md]: [97b049e](https://github.com/kontent-ai/delivery-sdk-net/pull/347/commits/97b049e5fd2ffb706af424c14093eb80465e0d1d)

File: docs/configuration/multiple-delivery-clients.md [hunk 7]
```diff
@@ -1,36 +1,128 @@
...
+The SDK provides extension methods upon the `IServiceCollection` that allow registering delivery client factory builder. The builder is used to configure the factory and register the delivery client instances.
+
+```csharp
+public class Startup
+{
+    // ...
+
+    public void ConfigureServices(IServiceCollection services)
+    {
+        services.AddDeliveryClientFactory
```
[the remark this answers is not in the document]
Comment (Simply007) [reply docs/configuration/multiple-delivery-clients.md]: [97b049e](https://github.com/kontent-ai/delivery-sdk-net/pull/347/commits/97b049e5fd2ffb706af424c14093eb80465e0d1d)

File: Kontent.Ai.Delivery.Extensions.DependencyInjection/INamedServiceProvider.cs [hunk 8]
```diff
@@ -1,8 +1,11 @@
-﻿namespace Kontent.Ai.Delivery.Extensions.DependencyInjection
+﻿using System;
+
+namespace Kontent.Ai.Delivery.Extensions.DependencyInjection
 {
     /// <summary>
     /// Represents a contract for retrieving named dependencies.
     /// </summary>
+    [Obsolete("#312")]
```
[the remark this answers is not in the document]
Comment (Simply007) [reply Kontent.Ai.Delivery.Extensions.DependencyInjection/INamedServiceProvider.cs:8]: 5ee8623

File: Kontent.Ai.Delivery.Extensions.DependencyInjection/NamedDeliveryClientFactory.cs [hunk 9]
```diff
@@ -16,7 +16,7 @@ internal class NamedDeliveryClientFactory : IDeliveryClientFactory
         private readonly ConcurrentDictionary<string, IDeliveryClient> _cache = new ConcurrentDictionary<string, IDeliveryClient>();
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="DeliveryClientFactory"/> class.
+        /// Initializes a new instance of the <see cref="MultipleDeliveryClientFactory"/> class.
```
[the remark this answers is not in the document]
Comment (Simply007) [reply Kontent.Ai.Delivery.Extensions.DependencyInjection/NamedDeliveryClientFactory.cs]: 0520331

File: Kontent.Ai.Delivery.Extensions.DependencyInjection/Extensions/ServiceCollectionExtensions.cs [hunk 10]
```diff
@@ -17,6 +19,26 @@ namespace Kontent.Ai.Delivery.Extensions.DependencyInjection
     /// </summary>
     public static class ServiceCollectionExtensions
```
[the remark this answers is not in the document]
Comment (Simply007) [reply Kontent.Ai.Delivery.Extensions.DependencyInjection/Extensions/ServiceCollectionExtensions.cs:20]: 4712164

File: Kontent.Ai.Delivery.Extensions.DependencyInjection/Extensions/ServiceCollectionExtensions.cs [hunk 11]
```diff
@@ -17,6 +19,26 @@ namespace Kontent.Ai.Delivery.Extensions.DependencyInjection
...
     {
+        /// <summary>
+        /// Registers a factory that will be used to configure multiple <see cref="IDeliveryClient"/> via <see cref="IDeliveryClientFactory"/>
+        /// </summary>
+        /// <param name="services">A <see cref="ServiceCollection"/> instance for registering and resolving dependencies.</param>
+        /// <param name="buildDeliveryClientFactory">A function that is provided with an instance of <see cref="IDeliveryClientFactoryBuilder"/>and expected to return a valid instance of <see cref="IDeliveryClientFactory"/>.</param>
```
[the remark this answers is not in the document]
Comment (Simply007) [reply Kontent.Ai.Delivery.Extensions.DependencyInjection/Extensions/ServiceCollectionExtensions.cs]: eb8bb2f

File: Kontent.Ai.Delivery.Extensions.DependencyInjection.Tests/ServiceCollectionEctensionsTests.MultipleDeliveryClientFactory.cs [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+            }
+
+            [Fact]
+            public void AddDeliveryClientFactory_NullOptions_ThrowsArgumentNullException()
+            {
+                Assert.Throws<ArgumentNullException>(() => _serviceCollection.AddMultipleDeliveryClientFactory(factoryBuilder => factoryBuilder.AddDeliveryClient(_correctName, null).Build()));
+            }
+
+            [Fact]
+            public void AddDeliveryClientFactory_NullName_ThrowsArgumentNullException()
```
[the remark this answers is not in the document]
Comment (Simply007) [reply Kontent.Ai.Delivery.Extensions.DependencyInjection.Tests/ServiceCollectionEctensionsTests.MultipleDeliveryClientFactory.cs:145]: aba7afb

File: Kontent.Ai.Delivery.Extensions.DependencyInjection.Tests/ServiceCollectionEctensionsTests.MultipleDeliveryClientFactory.cs [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+
+                var services = _serviceCollection.BuildServiceProvider();
+                var factory = services.GetRequiredService<IDeliveryClientFactory>();
+
+                var client = factory.Get(clientName);
+                client.Should().BeOfType<DeliveryClientCache>();
+            }
+
+            [Fact]
+            public void AddDeliveryClientFactory_NullOptions_ThrowsArgumentNullException()
```
[the remark this answers is not in the document]
Comment (Simply007) [reply Kontent.Ai.Delivery.Extensions.DependencyInjection.Tests/ServiceCollectionEctensionsTests.MultipleDeliveryClientFactory.cs:139]: aba7afb

File: Kontent.Ai.Delivery.Extensions.DependencyInjection.Tests/Builders/DeliveryClientDictionaryFactoryTests.cs [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+            var result = deliveryClientFactory.Get(_clientName);
+
+            result.Should().NotBeNull();
+        }
+
+        [Fact]
+        public void GetNamedClient_WithWrongName_RaiseArgumentException()
+        {
+            var fakeClient = A.Fake<IDeliveryClient>();
```
[the remark this answers is not in the document]
Comment (Simply007) [reply Kontent.Ai.Delivery.Extensions.DependencyInjection.Tests/Builders/DeliveryClientDictionaryFactoryTests.cs:44]: 32ba6bf

File: Kontent.Ai.Delivery.Extensions.DependencyInjection.Tests/Builders/DeliveryClientDictionaryFactoryTests.cs [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+        [Fact]
+        public void CreateFactory_NoClient_RaiseArgumentNullException()
+        {
+            Assert.Throws<ArgumentNullException>(() => new MultipleDeliveryClientFactory(null));
+        }
+
+        [Fact]
+        public void GetNamedClient_WithCorrectName_GetClient()
+        {
+            var fakeClient = A.Fake<IDeliveryClient>();
```
[the remark this answers is not in the document]
Comment (Simply007) [reply Kontent.Ai.Delivery.Extensions.DependencyInjection.Tests/Builders/DeliveryClientDictionaryFactoryTests.cs:33]: 32ba6bf

File: Kontent.Ai.Delivery.Extensions.DependencyInjection.Tests/DeliveryClientCacheFactoryTests.cs [hunk 16]
```diff
@@ -9,6 +10,7 @@
 
 namespace Kontent.Ai.Delivery.Extensions.DependencyInjection.Tests
 {
+    [Obsolete("#312")]
```
[the remark this answers is not in the document]
Comment (Simply007) [reply Kontent.Ai.Delivery.Extensions.DependencyInjection.Tests/DeliveryClientCacheFactoryTests.cs:13]: 0520331

File: Kontent.Ai.Delivery.Abstractions/Configuration/DeliveryOptions.cs [hunk 17]
```diff
@@ -75,6 +77,7 @@ public class DeliveryOptions
         /// <summary>
         /// The name of the service configuration this options object is related to.
         /// </summary>
+        [Obsolete("#312")]
```
[the remark this answers is not in the document]
Comment (Simply007) [reply Kontent.Ai.Delivery.Abstractions/Configuration/DeliveryOptions.cs:80]: 4712164

File: Kontent.Ai.Delivery.Caching/DeliveryCacheOptions.cs [hunk 18]
```diff
@@ -31,6 +31,7 @@ public class DeliveryCacheOptions
         /// <summary>
         /// The name of the service configuration this options object is related to.
         /// </summary>
+        [Obsolete("#312")]
```
[the remark this answers is not in the document]
Comment (Simply007) [reply Kontent.Ai.Delivery.Caching/DeliveryCacheOptions.cs:34]: 4712164

File: docs/configuration/multiple-delivery-clients.md [hunk 19]
```diff
@@ -1,36 +1,128 @@
...
+                .AddDeliveryClient
+                (
+                    "projectA",
+                    deliveryOptionBuilder => deliveryOptionBuilder
+                        .WithProjectId("<A_PROJECT_ID>")
+                        .UseProductionApi()
+                        .Build()
+                    optionalClientSetup =>
+                        optionalClientSetup.WithTypeProvider(new ProjectAProvider())
+                )
```
Comment (dzmitryk-kontent-ai) [inline docs/configuration/multiple-delivery-clients.md:42]: It seems that
`.Build()`
is missed here

File: docs/configuration/multiple-delivery-clients.md [hunk 20]
```diff
@@ -1,36 +1,128 @@
...
+                .AddDeliveryClient(
+                    "projectB",
+                    deliveryOptionBuilder => deliveryOptionBuilder
+                        .WithProjectId("<B_PROJECT_ID>")
+                        .UseProductionApi()
+                        .Build(),
+                    optionalClientSetup =>
+                        optionalClientSetup.WithTypeProvider(new ProjectBProvider())
+                )
```
Comment (dzmitryk-kontent-ai) [inline docs/configuration/multiple-delivery-clients.md:80]: It seems that
`.Build()`
is missed here

File: docs/configuration/multiple-delivery-clients.md [hunk 21]
```diff
@@ -1,36 +1,128 @@
...
+                            .Build(),
+                        CacheManagerFactory.Create(
+                            new MemoryCache(new MemoryCacheOptions()),
+                            Options.Create(new DeliveryCacheOptions
+                            {
+                                CacheType = CacheTypeEnum.Memory
+                            })),
+                        optionalClientSetup =>
+                            optionalClientSetup.WithTypeProvider(new ProjectAProvider())
+                    )
```
Comment (dzmitryk-kontent-ai) [inline docs/configuration/multiple-delivery-clients.md:117]: It seems that
`.Build()`
is missed here

File: docs/configuration/multiple-delivery-clients.md [hunk 22]
```diff
@@ -44,13 +136,30 @@ public class Startup
...
+                    _ =>
+                    {
+                        var options = new DeliveryOptions();
+                        config.Configuration.GetSection("MultipleDeliveryOptions:ProjectA")
+                            .Bind(options);
+                        return options;
+                    },
+                    optionalClientSetup =>
+                        optionalClientSetup.WithTypeProvider(new ProjectAProvider())
+                )
```
Comment (dzmitryk-kontent-ai) [inline docs/configuration/multiple-delivery-clients.md:154]: It seems that
`.Build()`
is missed here

**Review by dzmitryk-kontent-ai — CHANGES_REQUESTED** [review]

File: docs/configuration/multiple-delivery-clients.md (hunk 22 above)
Comment (Simply007) [reply docs/configuration/multiple-delivery-clients.md:157]: ae78d43

File: docs/configuration/multiple-delivery-clients.md (hunk 21 above)
Comment (Simply007) [reply docs/configuration/multiple-delivery-clients.md:119]: ae78d43

File: docs/configuration/multiple-delivery-clients.md (hunk 19 above)
Comment (Simply007) [reply docs/configuration/multiple-delivery-clients.md:42]: ae78d43

File: docs/configuration/multiple-delivery-clients.md (hunk 20 above)
Comment (Simply007) [reply docs/configuration/multiple-delivery-clients.md:81]: ae78d43

**Review by dzmitryk-kontent-ai — APPROVED** [review]

**Pull request merged**